### PR TITLE
GELF logger: Add gelf-compression-type and gelf-compression-level

### LINKED ISSUE #19665: Option to configure gelf compression
I think the Docker GELF log driver lacks a *log-opt* parameter to configure GELF compression.

Docker GELF logs are currently gziped.
Unfortunately, all GELF implementations don't support gzip logs.
For instance, heka GELF input only accepts plain-text gelf logs.

The go-gelf library which Docker relies on actually seems to support:
* CompressGzip
* CompressZlib
* CompressNone

### PR DESCRIPTION
**Pull request by dqminh**
This allows user to customize their compression level and also compression types of gelf messages.
- gelf-compression-level: from flate.DefautlCompression to flate.BestSpeed
- gelf-compression-type: gzip / zlib / none for disabled compression

cc @mariussturm

fixes #19665

### CODE REVIEW

**Comment by thaJeztah** [comment]
Don't forget to update the documentation as well

**Comment by calavera** [comment]
LGTM

File: daemon/logger/gelf/gelf.go [hunk 1]
```diff
@@ -107,7 +136,7 @@ func (s *gelfLogger) Log(msg *logger.Message) error {
 		Short:    string(msg.Line),
 		TimeUnix: float64(msg.Timestamp.UnixNano()/int64(time.Millisecond)) / 1000.0,
 		Level:    level,
-		Extra:    s.extra,
+		RawExtra: s.rawExtra,
```
Comment (bobrik) [inline daemon/logger/gelf/gelf.go]:
```json
{"version":"1.1","host":"36com10","short_message":"hell yeah","timestamp":1.454950493557e+09,"level":6"_command":"echo hell yeah","_container_id":"xxx","_container_name":"mad_mclean","_created":"2016-02-08T16:54:51.919800941Z","_image_id":"sha256:xxx","_image_name":"alpine:3.3","_tag":"cadvisor"}
```

Look what happens after `"level":6`

**Comment by bobrik** [comment]
This is blocked on https://github.com/Graylog2/go-gelf/pull/4.

**Comment by calavera** [comment]
Thanks for checking @bobrik.

**Comment by thaJeztah** [comment]
@bobrik @dqminh I see https://github.com/Graylog2/go-gelf/pull/4 was merged

**Comment by dqminh** [comment]
updated go-gelf
@thaJeztah i can add more docs for this if we think the design / flag is ok :)

**Comment by thaJeztah** [comment]
Hitting this again on gccgo; https://github.com/docker/docker/issues/19368 :cry:

**Comment by runcom** [comment]
LGTM moving to doc review

**Comment by thaJeztah** [comment]
I think the options need to be added to https://github.com/docker/docker/blob/master/docs/admin/logging/overview.md#gelf-options

(ideally, GELF should have its own page, like the other logging drivers, but we can do that in a separate PR)

Note that this will also need changes to the completion scripts, e.g. https://github.com/docker/docker/blob/master/contrib/completion/bash/docker#L412,  so let me ping @albers and @sdurrheimer

**Comment by tiborvass** [comment]
Ping @dqminh

**Comment by thaJeztah** [comment]
ping @dqminh can you address the docs comments then we can merge this

**Comment by ejholmes** [comment]
Ping as well! The poor performance of the gelf logging driver prevents it from being viable for production deployments ([Docker ends up soaking up all of the CPU GC'ing gelf buffers](https://dl.dropboxusercontent.com/u/1906634/Docker%20Benchmarks/docker-1.10.2%2Bgelf.pdf)). Hopefully the improvements made in https://github.com/Graylog2/go-gelf/pull/2 by @dqminh will help here!

**Comment by dqminh** [comment]
Sorry for the delay. Updated with docs.

PTAL @thaJeztah

File: daemon/logger/gelf/gelf.go [hunk 2]
```diff
@@ -127,15 +156,26 @@ func (s *gelfLogger) Name() string {
...
+			}
+		case "gelf-compression-type":
+			switch val {
+			case "gzip", "zlib", "none":
+			default:
+				return fmt.Errorf("unknown value %q for log opt %q for gelf log driver", val, key)
+			}
 		default:
-			return fmt.Errorf("unknown log opt '%s' for gelf log driver", key)
+			return fmt.Errorf("unknown log opt x%q for gelf log driver", key)
```
Comment (thaJeztah) [inline daemon/logger/gelf/gelf.go]: oh, looks like a stray `x` here

**Comment by thaJeztah** [comment]
Thank you @dqminh I spotted one typo in the error; also can you squash your last two commits? Docs LGTM otherwise

**Comment by vdemeester** [comment]
Docs LGTM 🐹

**Comment by thaJeztah** [comment]
Looks like azure-windows-tp4-4 needs a reboot; networking issues

**Pull request merged**